In [2]:
from sklearn.datasets import load_wine
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn import svm
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.metrics import accuracy_score


In [3]:
wine = load_wine() 
wine_data = wine.data 
wine_label = wine.target 

In [4]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [6]:
wine_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [7]:
data = wine_data
label = wine_label

X_train, X_test, y_train, y_test = train_test_split(data, #쓸 데이터(== feature)
                                                    label, #쓸 라벨(==target,==label)
                                                    test_size=0.2, #전체 데이터 중 '20%'를 test set에 사용하겠다!
                                                    random_state=7)


In [8]:
tree_model = DecisionTreeClassifier(random_state=10) 
#random_state는 train데이터와 test 데이터를 분리하는데 적용되는 랜덤성을 결정!
#라벨은 0부터 정렬이 되어있기 때문에 feature의 (label/target)이 동일해지기 때문
#하지만 random_state의 값을 일치시켜준다면 같은 랜덤성을 받을 수 있다!
#즉, 같은 train_data와 test_data를 가질 수 있다 !
tree_model.fit(X_train, y_train)
#모델 학습!
y_pred = tree_model.predict(X_test)
print(classification_report(y_test, y_pred))
accuracy = accuracy_score(y_test, y_pred)
print("-----------------------------------------------------------------")
print("Accuracy") 
accuracy

              precision    recall  f1-score   support

           0       0.88      1.00      0.93         7
           1       0.94      1.00      0.97        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.94      0.94      0.94        36
weighted avg       0.95      0.94      0.94        36

-----------------------------------------------------------------
Accuracy


0.9444444444444444

In [9]:
forest_model = RandomForestClassifier(random_state=10)
forest_model.fit(X_train, y_train)
y_pred = forest_model.predict(X_test)
print(classification_report(y_test, y_pred))
accuracy = accuracy_score(y_test, y_pred)
print("-----------------------------------------------------------------")
print("Accuracy") 
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

-----------------------------------------------------------------
Accuracy


1.0

이곳에서는 precision, recall, f1-score이 모두 1.00이 나와버렸다.  
다른 곳에서는 각각 다른 평가지표가 정상적이라고 생각되는 범주 안에서 나왔는데, 모델과 데이터가 맞지 않는건가 라는생각이 들었다.

In [10]:
sgd_model = SGDClassifier(random_state = 10)
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)
print(classification_report(y_test, y_pred))
accuracy = accuracy_score(y_test, y_pred)
print("-----------------------------------------------------------------")
print("Accuracy") 
accuracy

              precision    recall  f1-score   support

           0       0.44      1.00      0.61         7
           1       0.70      0.82      0.76        17
           2       0.00      0.00      0.00        12

    accuracy                           0.58        36
   macro avg       0.38      0.61      0.46        36
weighted avg       0.42      0.58      0.48        36

-----------------------------------------------------------------
Accuracy


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


0.5833333333333334

class 0에 대한 재현율이 1.00이 나왔다.

In [11]:
svm_model = svm.SVC(random_state=10);
svm_model.fit(X_train, y_train)
y_pred=svm_model.predict(X_test)
print(classification_report(y_test, y_pred))
accuracy = accuracy_score(y_test, y_pred)
print("-----------------------------------------------------------------")
print("Accuracy") 
accuracy

              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36

-----------------------------------------------------------------
Accuracy


0.6111111111111112

In [12]:
logistic_model = LogisticRegression(random_state=10)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)
print(classification_report(y_test, y_pred))
accuracy = accuracy_score(y_test, y_pred)
print("-----------------------------------------------------------------")
print("Accuracy") 
accuracy

              precision    recall  f1-score   support

           0       1.00      0.86      0.92         7
           1       0.94      1.00      0.97        17
           2       1.00      1.00      1.00        12

    accuracy                           0.97        36
   macro avg       0.98      0.95      0.96        36
weighted avg       0.97      0.97      0.97        36

-----------------------------------------------------------------
Accuracy


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.9722222222222222

이곳은 sgd와 다르게 class 0과 2에 대해서 precision이 1, 1과 2에 대해서 recall이 1, class 2에 대해서 f1-score가 1이 나왔다.  
  
forest_model에서 accuracy 기준 1.0, logistic regression에서 0.97이 나왔다.

### Accuracy  
와인 분류기 또한 손글씨 분류기와 같이 Accuracy를 평가지표로 택했다.  
와인 분류기의 타겟(label)은 class 0, class1, class2가 존재했는데,(class 별 각 의미가 명확하지 않아서 품종으로 생각했다!)  
이또한 이진법적으로 양성이냐, 음성이냐의 차이가 아니라 wine1(예시)이 가지고 있는 속성(말산, 도수 등)을 통해서 원하는 품종으로 나누어야 했기 때문이다
